pkg/server/ui/ui.go

```diff
@@ -1,29 +1,132 @@
+	VueIndex = Vue.IndexFile()
+)
 
-func JSURLGetter() string {
-	if settings.UIIndex.Get() == "local" {
-		return "/api-ui/ui.min.js"
+func newHandler(
+	indexSetting func() string,
+	pathSetting func() string,
+	offlineSetting func() string) *handler {
+	return &handler{
+		indexSetting:   indexSetting,
+		offlineSetting: offlineSetting,
+		pathSetting:    pathSetting,
+		middleware: responsewriter.Chain{
+			responsewriter.Gzip,
+			responsewriter.DenyFrameOptions,
+			responsewriter.CacheMiddleware("json", "js", "css"),
+		}.Handler,
+		indexMiddleware: responsewriter.Chain{
+			responsewriter.Gzip,
+			responsewriter.NoCache,
+			responsewriter.DenyFrameOptions,
+			responsewriter.ContentType,
+		}.Handler,
 	}
-	return ""
 }
 
-func CSSURLGetter() string {
-	if settings.UIIndex.Get() == "local" {
-		return "/api-ui/ui.min.css"
+type handler struct {
+	pathSetting     func() string
+	indexSetting    func() string
+	offlineSetting  func() string
+	middleware      func(http.Handler) http.Handler
+	indexMiddleware func(http.Handler) http.Handler
+
+	downloadOnce    sync.Once
+	downloadSuccess bool
+}
+
+func (u *handler) canDownload(url string) bool {
+	u.downloadOnce.Do(func() {
+		if err := serveIndex(ioutil.Discard, url); err == nil {
+			u.downloadSuccess = true
+		} else {
+			logrus.Errorf("Failed to download %s, falling back to packaged UI", url)
+		}
+	})
+	return u.downloadSuccess
+}
+
+func (u *handler) path() (path string, isURL bool) {
+	switch u.offlineSetting() {
+	case "dynamic":
+		if settings.IsRelease() {
+			return u.pathSetting(), false
+		}
+		if u.canDownload(u.indexSetting()) {
+			return u.indexSetting(), true
+		}
+		return u.pathSetting(), false
+	case "true":
+		return u.pathSetting(), false
+	default:
+		return u.indexSetting(), true
 	}
-	return ""
 }
 
-func APIUIVersionGetter() string {
-	return settings.APIUIVersion.Get()
+func (u *handler) ServeAsset() http.Handler {
+	return u.middleware(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		http.FileServer(http.Dir(u.pathSetting())).ServeHTTP(rw, req)
+	}))
+}
+
+func (u *handler) IndexFileOnNotFound() http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		req.URL.Path = strings.TrimPrefix(req.URL.Path, "/dashboard")
+		if _, err := os.Stat(filepath.Join(u.pathSetting(), req.URL.Path)); err == nil {
+			u.ServeAsset().ServeHTTP(rw, req)
+		} else {
+			u.IndexFile().ServeHTTP(rw, req)
+		}
+	})
+}
+
+func (u *handler) IndexFile() http.Handler {
+	return u.indexMiddleware(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		if path, isURL := u.path(); isURL {
+			_ = serveIndex(rw, path)
+		} else {
+			http.ServeFile(rw, req, filepath.Join(path, "index.html"))
+		}
+	}))
+}
+
+func serveIndex(resp io.Writer, url string) error {
+	r, err := insecureClient.Get(url)
+	if err != nil {
+		return err
+	}
+	defer r.Body.Close()
+
+	_, err = io.Copy(resp, r.Body)
+	return err
 }
```

Comment:
Can we serve the content as we did in apiserver? And why do we need a plain text request?

Reply:
steve API's dashboard UI has been removed in https://github.com/rancher/steve/commit/a3fbe499d1cc570043ff1385d6168b0c6a768df7#diff-616c650128f9d441dd01032ab9d8fec275b8c517a7d852f775d58b6f5ecf1a88
therefore added serve UI index in the Harvester by following the Rancher upstream.